### PR TITLE
feat(state-tree-depth) - ensure stateTreeDepth is not hardcoded 

### PR DESCRIPTION
**Pull request by ctrlc03**
We do not want to hardcode the stateTreeDepth in the smart contracts, but instead we want to pass them as arguments.

Changes:

* in the smart contract this is an immutable param thus passed in the constructor
* MaciState.ts and Poll.ts take them as constructor argument
* in tests we hardcode to 10 for the sake of testing
* contracts package.json's script `compileSol` takes an additional parameter so that we can compile the trees contracts with the correct params

### CODE REVIEW

File: contracts/contracts/MACI.sol [hunk 1]
```diff
@@ -129,6 +131,7 @@ contract MACI is IMACI, DomainObjs, Params, SnarkCommon, Ownable {
         pollFactory = _pollFactory;
         signUpGatekeeper = _signUpGatekeeper;
         initialVoiceCreditProxy = _initialVoiceCreditProxy;
+        stateTreeDepth = _stateTreeDepth;
```
Comment (samajammin) [inline contracts/contracts/MACI.sol:134]: Is there any danger in this value now being an arbitrary size? i.e. should there be a check against some max value?

File: contracts/ts/__tests__/MACI.test.ts [hunk 2]
```diff
@@ -113,12 +112,15 @@ describe('MACI', () => {
...
     let pollId: number
+    let user: ethers.Signer
 
     describe('Deployment', () => {
         beforeAll(async () => {
             signer = await getDefaultSigner()
             const r = await deployTestContracts(
                 initialVoiceCreditBalance,
+                STATE_TREE_DEPTH,
+                true
```
Comment (samajammin) [inline contracts/ts/__tests__/MACI.test.ts:123]: Why pass `true` here?

Comment (ctrlc03) [reply contracts/ts/__tests__/MACI.test.ts:123]: oh basically those utils functions like deployTestContract or deployContract have an optional parameter called "quiet", passing true so it doesn't print unnecessary logs while running the tests

File: contracts/contracts/MACI.sol (hunk 1 above)
Comment (ctrlc03) [reply contracts/contracts/MACI.sol:134]: Yes and no. If they put a value that does not match the circuit, they would basically need to start over (there might be some workarounds though, still not ideal).

This is a uint8 so max value would be 255, having said that, a tree depth of 255 would be incredibly big, considering 10 allows for a little bit more than 1M total signups. 
My view here is the following:
1. if we put a max value, what should this be?
2. In the past, we asked the users to change a constant variable rather than passing it as a constructor parameter:
   - I consider these options kinda the same, we never really validated that this constant variable was > than a max value
   - people running MACI should be (and are to some extent) informed that this parameter should match the one in the circuits, hence it is their responsibility of setting it up correctly.
3. even if this value was < than a max value we decide on, it could still be different than what they put in the circuit and in the other components of MACI. 

To summarize, I think the best we can do here in terms of "validation" is to properly inform that this value should match whichever value is in the Circom circuits.

**Comment by baumstern** [comment]
To confirm, we don't have to follow the steps described in [this doc](https://github.com/privacy-scaling-explorations/maci/wiki/Follow-ups-when-changing-%60stateTreeDepth%60 ) anymore?

**Comment by ctrlc03** [comment]
> To confirm, we don't have to follow the steps described in [this doc](https://github.com/privacy-scaling-explorations/maci/wiki/Follow-ups-when-changing-%60stateTreeDepth%60) anymore?

correct, I would say we probably need to improve documentation on this change - for now I updated the scripts in the contracts folder, as well as the README. What do you think, should we updated the cli docs too to ensure the new flag ("-s") is more evident? would be curious to hear your opinion after you review this PR and how it worked for you testing it

File: circuits/ts/__tests__/utils.ts [hunk 3]
```diff
@@ -26,7 +26,4 @@ async function getSignal(circuit: CircuitT, witness: bigint[], name: string): Pr
     return BigInt(witness[indexInWitness]);
 }
 
-export {
-    str2BigInt,
-    getSignal
-}
+export const STATE_TREE_DEPTH = 10
```
Comment (baumstern) [inline circuits/ts/__tests__/utils.ts:29]: Is there a specific reason this is here? Perhaps it could fit well in `constants.ts`

File: crypto/ts/__tests__/AccQueue.test.ts [hunk 4]
```diff
@@ -346,7 +346,6 @@ describe('AccQueue', () => {
                 })
 
                 it('Should fail if there are 0 leaves', () => {
-                    expect.assertions(1)
```
Comment (baumstern) [inline crypto/ts/__tests__/AccQueue.test.ts:349]: Just curious, why was this removed?

File: crypto/ts/__tests__/Crypto.test.ts [hunk 5]
```diff
@@ -111,7 +111,6 @@ describe('Cryptographic operations', () => {
         })
 
         it('The ciphertext should differ from the plaintext', () => {
-            expect.assertions(plaintext.length)
```
Comment (baumstern) [inline crypto/ts/__tests__/Crypto.test.ts:114]: Just curious, why was this removed?

**Comment by baumstern** [comment]
> > To confirm, we don't have to follow the steps described in [this doc](https://github.com/privacy-scaling-explorations/maci/wiki/Follow-ups-when-changing-%60stateTreeDepth%60) anymore?
> 
> correct, I would say we probably need to improve documentation on this change - for now I updated the scripts in the contracts folder, as well as the README. What do you think, should we updated the cli docs too to ensure the new flag ("-s") is more evident? would be curious to hear your opinion after you review this PR and how it worked for you testing it

It’s an essential value, so it should be as visible as possible. However, people often don’t read the documentation thoroughly..so we might need to seek a somewhat better way to highlight it. Although a good idea doesn’t come to mind immediately.

File: circuits/ts/__tests__/utils.ts (hunk 3 above)
Comment (ctrlc03) [reply circuits/ts/__tests__/utils.ts:29]: totally

File: crypto/ts/__tests__/AccQueue.test.ts (hunk 4 above)
Comment (ctrlc03) [reply crypto/ts/__tests__/AccQueue.test.ts:349]: I'm not really sure these are needed tbf. You can use expect.assertions when there might be cases of certain assertions are not called, and you want to constraint it to actually check those run. In this case it seems random to me.. For instance:

```ts 
   expect.assertions(plaintext.length)
   for (let i = 0; i < plaintext.length; i++) {
                expect(plaintext[i] !== ciphertext[i + 1]).toBeTruthy()
    }
```
we are looping for plaintext.length iterations, so it's kinda obvious that it will call expect for plaintext.length times

On the other hand, if you had this (very simple and dumb example): 

```ts
 const data = fetchData();
  if (data) {
        expect(data).toBe('expected data');
 }
```
here it would make sense to add expect.assertions(1) -> if you are running the test cases assuming that `data` should always be defined

File: crypto/ts/__tests__/Crypto.test.ts (hunk 5 above)
Comment (ctrlc03) [reply crypto/ts/__tests__/Crypto.test.ts:114]: read above please

File: crypto/ts/__tests__/AccQueue.test.ts (hunk 4 above)
Comment (baumstern) [reply crypto/ts/__tests__/AccQueue.test.ts:349]: thanks for the explaination! Well as you said there is no meaning of using `expect.assertion()` in this code block.

Comment (baumstern) [reply crypto/ts/__tests__/AccQueue.test.ts:349]: Anyway we don't use jest anymore so it doesn't matter at all

**Comment by baumstern** [comment]
Just a reminder that there is a merge conflict that needs to be addressed. Once that's resolved, it looks like we're all set to merge this PR into the dev branch. Great work!

**Comment by ctrlc03** [comment]
> Just a reminder that there is a merge conflict that needs to be addressed. Once that's resolved, it looks like we're all set to merge this PR into the dev branch. Great work!

merge conflicts sorted!  thanks for the review

**Review by baumstern — APPROVED** [review]
LGTM
